src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -20,86 +25,118 @@ class BankManager {
+                                  const std::shared_ptr<const ov::ICore>& core,
+                                  const std::string& alloc_device);
 
 private:
     // Data
     std::unordered_map<std::string, std::weak_ptr<Bank>> m_bank_map;
     std::mutex m_mutex;
 };
 
-ov::Tensor Bank::update(const ov::Tensor& tensor) {
-    if (!tensor) {
-        OPENVINO_THROW("Uninitialized tensor in weights bank allocation!");
-    }
+ov::Tensor Bank::get(const LazyTensor& tensor, const std::string& device) {
+    const std::string& device_for_alloc = m_alloc_device.empty() ? device : m_alloc_device;
 
     std::lock_guard<std::mutex> guard(m_mutex);
 
-    if (m_bank.find(tensor.data()) == m_bank.end()) {
-        // need to allocate first
-        m_bank[tensor.data()] = tensor;
+    auto& device_bank = m_device_bank[device_for_alloc];
+    auto iter_device = device_bank.find(tensor);
+
+    if (iter_device != device_bank.end() && iter_device->second) {
+        // Already allocated
+        return iter_device->second;
     }
 
-    return tensor;
+    // Allocation and evaluation needed
+    return unsafe_eval_and_alloc(tensor, device_for_alloc);
 }
 
-ov::Tensor Bank::get(const ov::Tensor& tensor, const std::string& device) {
-    if (!tensor) {
-        OPENVINO_THROW("Uninitialized tensor in weights bank allocation!");
-    }
-
-    if (device != "CPU" && device != "NPU") {
-        OPENVINO_THROW("Unsupported device in weights bank allocation: ", device);
-    }
+void Bank::registerLT(const LazyTensor& tensor, const std::string& device) {
+    const std::string& device_for_alloc = m_alloc_device.empty() ? device : m_alloc_device;
 
     std::lock_guard<std::mutex> guard(m_mutex);
 
-    auto iter_cpu = m_bank.find(tensor.data());
-    if (iter_cpu == m_bank.end()) {
-        OPENVINO_THROW("Unknown tensor in weights bank allocation!");
+    auto& device_bank = m_device_bank[device_for_alloc];
+    if (device_bank.find(tensor) == device_bank.end()) {
+        device_bank[tensor] = ov::Tensor();
     }
+}
 
-    // If target device is CPU - just reuse the default bank
-    if (device == "CPU") {
-        return iter_cpu->second;
+void Bank::evaluate_and_allocate() {
+    std::lock_guard<std::mutex> guard(m_mutex);
+
+    for (auto&& bank : m_device_bank) {
+        const auto& device_for_alloc = bank.first;
+        auto& device_bank = bank.second;
+        std::vector<LazyTensor> vec;
+        for (const auto& el : device_bank) {
+            vec.push_back(el.first);
+        }
+        // FIXME: this is supposed to be parallel_for, however L0 mutex needs to be aquired for allocation
+        for (std::size_t idx = 0; idx < vec.size(); ++idx) {
+            const auto& lt = vec[idx];
+            auto iter_device = device_bank.find(lt);
+            if (iter_device != device_bank.end() && iter_device->second) {
+                // Already allocated
+                return;
+            }
```

Comment:
Bug here. It was a parallel_for, now it is a for - so your `return` exits the whole `evaluate_and_allocate()`, not the lambda (loop body) which was there before.

Reply:
So better add mutex here. :)

Reply:
Thanks for catching it! Fixed